### PR TITLE
VCB-17 Debug Step Mode

### PR DESCRIPTION
**Pull request by clooudch**

### CODE REVIEW

File: src/assembler/index.ts [hunk 1]
```diff
@@ -42,6 +42,11 @@ function createExecutable(objectFile: IELF): IELF {
...
       ...objectFile.content,
       ...END_OF_CODE.toBytes()
-    ]
+    ],
+    sourceMap: new Map<number, number>()
   }
+  objectFile.sourceMap.forEach((line, offset) => {
+    executable.sourceMap.set(FLASH_START.add(offset + 8).value, line)
+  })
+  return executable
```
Comment (deflox) [inline src/assembler/index.ts:51]: Maybe I do not see the whole picture but why do we again create an IELF object from the existing IELF object within the `createExecutable()` function? As far as I can see, the `vectorTable` and the `END_OF_CODE` are added. Is this just for better separation? Or couldn't the whole object straight away be build in the encoder module?

File: src/assembler/parser/index.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * @returns the parsed code file (AST representation)
+ */
+export function parse(code: string): ICodeFile {
+  const ast: ICodeFile = {
+    symbols: {},
+    areas: []
+  }
+  let area: IArea | null = null
+  let label: string | null = null
+  const rules: ITextParseRule[] = [
```
Comment (deflox) [inline src/assembler/parser/index.ts:19]: Really cool structure of first defining the rules and afterwards let the parser do the job with the given array of rules. ^^

File: src/assembler/parser/text.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+/**
+ * Parses text using the specified rules.
+ *
+ * @param text the text to be parsed
+ * @param rules the rules to be used for parsing
+ * @returns the last cursor position before there was no more match
+ */
+export function parseText(text: string, rules: ITextParseRule[]): ITextCursor {
```
Comment (deflox) [inline src/assembler/parser/text.ts:36]: Maybe it would make sense to also add dedicated tests for this module? But perhaps it is already more than enough with covered with the parser tests.

**Review by deflox — COMMENTED** [review]
Here's my feedback. Overall looks good.

File: src/assembler/ast.ts [hunk 4]
```diff
@@ -24,11 +31,13 @@ export interface IInstruction {
...
   label?: string
   options: string[]
+  line: number
 }
 
 /**
  * Represents the code of an assembler file.
  */
 export interface ICodeFile {
+  symbols: ISymbols
```
[the remark this answers is not in the document]
Comment (clooudch) [reply src/assembler/ast.ts:41]: I do not think that symbols declarations (using `EQU`) have any special meaning when defined in an area. Symbols are not really part of a section but rather a global entry in the symbol table.

I would suggest to just keep it like it is right now. -> We collect all symbol definitions and just store them globally in the symbol table. There will be no error thrown when symbol definitions are within an area, they can be anywhere.

[the remark this answers is not in the document]
Comment (clooudch) [reply src/assembler/ast.ts:41]: I am resolving this conversation. Feel free to re-open.

File: src/assembler/index.ts (hunk 1 above)
Comment (clooudch) [reply src/assembler/index.ts:51]: It's just for better separation. And also I think it will make it easier when we are adding more features in the future.

ELF stands for executable and linkable format. So, basically it's a special format which can be used for linking and for executing. The `assembler` (in our case the `encoder.ts` file) has the job to take a code file and convert it into an *linkable* ELF file. So, it does have sections and symbol table and everything you would need for the linking process. On the other hand, the `linker` (in our case the `assembler/index.ts` file) has the job to convert it into an executable ELF file by adding a program header (segment information) which tells the loader how this should be loaded and executed.

Comment (clooudch) [reply src/assembler/index.ts:51]: I will close this conversation. If you think you spotted an issue, do not hesitate to re-open it.

File: src/assembler/parser/text.ts (hunk 3 above)
Comment (clooudch) [reply src/assembler/parser/text.ts:36]: Inside of the parser tests, there are also tests for the text parser. I've created one test file for the whole parser folder.

File: src/assembler/parser/index.ts (hunk 2 above)
Comment (clooudch) [reply src/assembler/parser/index.ts:19]: Thanks :blush:

I had to come up with an alternative to the `lezer` parser system :sweat_smile:. It's a quite simple regex based text parser.

File: docs/editor-modes.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  * With the halt button the execution is halted and the editor moves into STEP mode.
+* STEP Mode:
+  * The currently compiled code execution is haltet at a certain point. As in RUN Mode, edits to the code are also not possible.
+  * With the reset button the board will be reset and the editor is moved into EDIT mode.
+  * With the step button the next instruction can be executed. The editor will stay in STEP mode.
+  * With the continue button the execution can be continued in RUN mode.
+
+Below is a state diagram that visualizes the three modes:
+
+![Editor Modes Diagram](assets/editor-modes.svg)
```
[the remark this answers is not in the document]
Comment (clooudch) [reply docs/editor-modes.md:20]: Thanks for spotting this. I fixed it.

**Comment by clooudch** [comment]
> Also at least locally there are still tests failures on my machine but this might be due to master not being merged yet:
> 
> image

Thanks for spotting this. I have fixed it. I somehow forgot to re-ran the tests after some final changes.

**Review by deflox — APPROVED** [review]
Since every remark from me has been resolved I approve from my side.

File: src/assembler/parser/index.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+          instructions: []
+        }
+        ast.areas.push(area)
+        label = null
+      }
+    },
+    {
+      name: 'Label',
+      pattern:
+        /([a-z0-9]+)(?=\s+([a-z]+) +([0-9a-z#\[\]]+(, *[0-9a-z#\[\]]+)))/,
```
Comment (m-dot-wigger) [inline src/assembler/parser/index.ts:55]: Can a label start with a digit?

File: src/assembler/parser/index.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      name: 'Whitespace',
+      pattern: /\s+/
+    },
+    {
+      name: 'Comment',
+      pattern: /;[^\n]*/
+    },
+    {
+      name: 'LiteralSymbolDeclaration',
+      pattern: /([a-z\|\._]+) +EQU +([0-9a-z#]+)/,
```
[the remark this answers is not in the document]
Comment (m-dot-wigger) [reply src/assembler/parser/index.ts:29]: found case-insensitive option in parser/text.ts

File: test/assembler/parser.test.ts [hunk 8]
```diff
@@ -1,158 +1,264 @@
...
-import { createInstruction, parse, removeNonCode } from 'assembler/parser'
+import { AreaType } from 'assembler/ast'
+import { parse } from 'assembler/parser'
+import { ITextParseRule, parseText } from 'assembler/parser/text'
 
-//// create test data
+const code = `
+MY_CONSTANT EQU 0x123
+AREA |.data|, DATA, READWRITE
+  DCD 0xFF ; A comment
```
Comment (m-dot-wigger) [inline test/assembler/parser.test.ts:8]: I am missing tests for comments on own line and empty lines without anything on it

**Review by m-dot-wigger — CHANGES_REQUESTED** [review]

**Review by ni507 — APPROVED** [review]

File: src/assembler/parser/index.ts (hunk 6 above)
Comment (m-dot-wigger) [reply src/assembler/parser/index.ts:55]: underscore is valid according to keil uvision but digits are not. See:
grafik

Comment (deflox) [reply src/assembler/parser/index.ts:55]: I have found this documentation from ARM Keil which states:
```
symbol must begin in the first column. It cannot contain any white space character such as a space or a tab unless it is enclosed by bars (|).
Labels are symbolic representations of addresses. You can use labels to mark specific addresses that you want to refer to from other parts of the code. Numeric local labels are a subclass of labels that begin with a number in the range 0-99. Unlike other labels, a numeric local label can be defined many times. This makes them useful when generating labels with a macro.
```
See https://www.keil.com/support/man/docs/ARMASM/armasm_dom1359731141352.htm

There is also a description about the label that is used for an area declaration:
```
You can choose any name for your sections. However, names starting with a non-alphabetic character must be enclosed in bars or a missing section name error is generated. For example, |1_DataArea|.
```

See: https://www.keil.com/support/man/docs/ARMASM/ARMASM_dom1361290002714.htm

So according to this, it should be possible if you enclose the label in bars ||.

Comment (deflox) [reply src/assembler/parser/index.ts:55]: But maybe this is for another compiler because I just tried it and it does not really work in our Keil uVision so I guess only alphanumeric characters are allowed then...

Comment (clooudch) [reply src/assembler/parser/index.ts]: I've updated the parser with the latest regex.

File: test/assembler/parser.test.ts (hunk 8 above)
Comment (clooudch) [reply test/assembler/parser.test.ts:9]: I extended the tests.

**Review by m-dot-wigger — APPROVED** [review]
like the way you inserted variables in Regular Expression
